design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+Cluster’s connection details - it uses them as Provider credentials. If the EKS
+Cluster is deleted before the Helm Release the Helm Release controller becomes
+unable to connect to its Provider in order to delete the Release. In practice
+the Release External Resource (ER) was implicitly deleted along with the GKE
+Cluster - there’s actually no work left for the Helm release controller to do -
+but it cannot know that. This results in the Helm Release MR being ‘orphaned’
+in a pending deletion that can never be satisfied.
+
+- **Orphaned ‘Side-Effect’ Resources.** Consider the same example as above, and
+assume the Helm Release includes a Kubernetes Service resource of type:
+LoadBalancer. A side-effect of the creation of this Service will be to create an
+ELB in the EKS cluster’s VPC. This ELB is not managed by or even known to
+Crossplane, and is not deleted along with the EKS cluster. This means that the
+Helm Release (and therefore its Service) must be deleted before the EKS cluster
+in order to ensure the EKS cluster’s controllers trigger deletion of the ELB.
+If it is not, the ELB will be orphaned, costing money and blocking the deletion
+of the VPC it was created in.
+
+These problems are exacerbated by the fact that Kubernetes uses "background
+cascading deletion" by default. Put otherwise, the successful deletion of a
+parent resource is not dependent on the successful deletion of its children.
+Instead, the parent resource will be deleted successfully and the children
+(i.e. resources that declare the parent as an owner reference) will later be
+deleted by the Kubernetes garbage collector. This makes it difficult to diagnose
+or even notice resources that become orphaned when their parent XR is deleted.
+
+## Goals
+
+- Support "cross-provider" dependencies, for example an MR from provider-helm
+depending on a ProviderConfig from provider-aws.
+- Don’t require Crossplane users to manage fine-grained RBAC access.
+- Don’t grant Crossplane providers broad RBAC access (e.g. to all other providers).
+
+## Proposal
+
+We propose introducing a new `Usage` type that can be used to declare usage
+relationships between Crossplane resources. This type will be defined by the
+Core Crossplane repository, and be available for use as part of a composition.
+
+The relations defined by this type will be enforced by an admission webhook that
+will be running as part of the Crossplane core. Similar to the upstreams [liens
+proposal], the webhook will reject deletions of resources that are in use by
+other resources with a ["409 Conflict" error code].
+
+### API
+
+We will introduce a new `Usage` type as follows:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Usage
+metadata:
+  name: release-uses-cluster
+spec:
+  of:
+  - apiVersion: eks.upbound.io/v1beta1
+    kind: Cluster
+    name: my-cluster
+  by:
+    apiVersion: helm.crossplane.io/v1beta1
+    kind: Release
+    name: my-prometheus-chart
```

Comment:
Should `by` be optional?

I'm thinking folks might want to manually create a `Usage` as a form of deletion protection. For example to add a layer of protection that a `VPC` not be accidentally deleted.

Reply:
Yeah I could be convinced the use case I describe should be out of scope.

If we did keep it in scope, maybe a `type` field would help:

```yaml
apiVersion: crossplane.io/v1
kind: Usage
spec:
  type: Protection  # Protection vs Dependency? Human vs Resource?
  of:
  - apiVersion: ec2.upbound.io/v1beta1
    kind: VPC
    name: very-critical-vpc
```

This could also be used for CEL validation - i.e. make `by` required if the `type` indicates it should be.

Reply:
Introducing `type` would help to figure out whether `by` should be required or not makes sense, however, does not still help with "enforcing documenting the source of the `Usage`". I am wondering if we can have a string `reason` field and require either `spec.by` or `spec.reason` with CEL rules? TBH, `spec.type` would be redundant in that case even I believe it is a convention but not sure applies here.

Protection:

```
apiVersion: crossplane.io/v1
kind: Usage
spec:
  reason: "Production Database - should never be deleted"
  of:
    apiVersion: rds.aws.upbound.io/v1beta1
    kind: Instance
    resourceRef:
      name: my-cluster
```

Dependency:

```
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Usage
metadata:
  name: release-uses-cluster
spec:
  of:
    apiVersion: eks.upbound.io/v1beta1
    kind: Cluster
    resourceRef:
      name: my-cluster
  by:
    apiVersion: helm.crossplane.io/v1beta1
    kind: Release
    resourceRef:
      name: my-prometheus-chart
```